compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -1552,12 +1576,72 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 }
 
                 err.emit_unless(unsized_return);
+                if ret_vec.len() > 0 {
+                    if let Some(expr) = expression {
+                        let mut addtion_err = fcx.tcx.sess.diagnostic().span_note_diag(expr.span, "the function expects a value to always be returned, but loops might run zero times");
+                        for i in 0..ret_vec.len() {
```

Comment:
what do you mean by this? Sounds like a new issue for me.
 Can you give me more details that we should emit?

Reply:
He means something like:

```
note: the function expects a value to always be returned, but loops might run zero times
  --> $DIR/issue-98982.rs:2:5
   |
LL |     for i in 0..0 {
   |     ^^^^^^^^^^^^^ this might have zero elements to iterate on
LL |
LL |         return i;
   |         -------- if the loop doesn't execute, this value would never get returned
...
LL |         return i;
   |         -------- if the loop doesn't execute, this value would never get returned
...
LL |         return i;
   |         -------- if the loop doesn't execute, this value would never get returned
note: if the loop doesn't execute, 5 other values would never get returned
```

Reply:
created https://github.com/rust-lang/rust/issues/100285 and assigned to myself